totalRP3/modules/NamePlates/NamePlates.lua

```diff
@@ -0,0 +1,1124 @@
+	if not name or name == "" or name == UNKNOWNOBJECT then
+		-- Don't return profiles for invalid/unknown units.
+		return nil;
+	end
+
+	return Player.CreateFromNameAndRealm(name, realm);
+end
+
+-- Returns the (combat) pet companion profile associated with the given
+-- unit token.
+--
+-- If no profile can be found, nil is returned.
+local function GetCombatPetProfile(unitToken)
+	local companionType = TRP3_UI.misc.TYPE_PET;
+	local fullID = TRP3_UI.misc.getCompanionFullID(unitToken, companionType);
+	if not fullID then
+		return nil;
+	end
+
+	local profile = TRP3_Companions.register.getCompanionProfile(fullID);
+	if not profile then
+		return nil;
+	end
+
+	return profile.data;
+end
+
+-- Returns an appropriate default OOC indicator style for configuration.
+local function GetDefaultOOCIndicator()
+	local defaultOOCIndicator = "TEXT";
+
+	-- The default OOC icon should inherit from the tooltip setting if
+	-- available, but without adding a module dependency please.
+	if TRP3_Configuration then
+		local tooltipOOCIndicator = TRP3_Configuration["tooltip_prefere_ooc_icon"];
+		if tooltipOOCIndicator == "TEXT" or tooltipOOCIndicator == "ICON" then
+			defaultOOCIndicator = tooltipOOCIndicator;
+		end
+	end
+
+	return defaultOOCIndicator;
+end
+
+-- Nameplate module table.
+local TRP3_NamePlates = {};
+
+-- Handler called when the module is being initialized. This is responsible
+-- for setting up the base internal state without actually enabling any
+-- behaviours.
+function TRP3_NamePlates:OnInitialize()
+	self.activeUnitTokens = {};
+	self.registerUnitIDMap = {};
+	self.registerUnitIDCooldowns = {};
+	self.unitFrameWidgets = {};
+
+	self.fontStringPool = CreateFontStringPool(UIParent, "ARTWORK", 0, "SystemFont_NamePlate");
+	self.texturePool = CreateTexturePool(UIParent, "ARTWORK", 0);
+
+	-- Register configuration keys. Do this on-init so the keys always exist.
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_ENABLE_CUSTOMIZATIONS, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_ONLY_IN_CHARACTER, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_ACTIVE_QUERY, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_SHOW_PLAYER_NAMES, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_SHOW_PET_NAMES, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_SHOW_COLORS, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_SHOW_ICONS, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_SHOW_TITLES, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_SHOW_OOC_INDICATORS, true);
+	TRP3_Config.registerConfigKey(CONFIG_NAMEPLATES_OOC_INDICATOR, GetDefaultOOCIndicator());
+
+	-- Prevent the module from actually being *enabled* if there's any of the
+	-- conflicting addons installed.
+	for _, addonName in ipairs(CONFLICTING_ADDONS) do
+		local state = GetAddOnEnableState(TRP3_Globals.player, addonName);
+		if state == 2 then
+			return false, format(L.NAMEPLATES_ERR_ADDON_CONFLICT, addonName);
+		end
+	end
```

Comment:
Could this be moved higher up, so we don't register config keys for nothing?

Reply:
Is there a reason to _not_ register config keys? No UI is added unless the module actually starts up, but one of the annoyances as noted with the tooltip settings is the fact that config keys will error if they're queried but not yet registered.

I'd rather have the keys registered all the time at a global level somewhere else honestly, for _all_ the settings. That way we don't need to paste them around the whole shop and run into issues where module A is disabled but it registers a key that module B might be interested in.

Reply:
Okay, that makes sense